Src/Common/CorelationIdLookupHelper.cs

```diff
@@ -0,0 +1,80 @@
+﻿namespace Microsoft.ApplicationInsights.Common
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Globalization;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// A store for instrumentation App Ids. This makes sure we don't query the public endpoint to find an app Id for the same Ikey more than once.
+    /// </summary>
+    internal static class CorelationIdLookupHelper
+    {
+        /// <summary>
+        /// Max number of component hashes to cache.
+        /// </summary>
+        private const int MAXSIZE = 100;
+
+        private const string CorrelationIdFormat = "aid-v1:{0}";
+
+        private static ConcurrentDictionary<string, string> knownAppIds = new ConcurrentDictionary<string, string>();
+
+        /// <summary>
+        /// Retrieves the hash for a given instrumentation key.
+        /// </summary>
+        /// <param name="instrumentationKey">Instrumentation key string.</param>
+        /// <param name="correlationId">AppId corresponding to the provided instrumentation key</param>
+        /// <returns>true if correlationId was successfully retrieved; false otherwise.</returns>
+        public static bool TryGetAppId(string instrumentationKey, out string correlationId)
+        {
+            if (string.IsNullOrEmpty(instrumentationKey))
+            {
+                throw new ArgumentNullException("instrumentationKey");
+            }
+
+            var found = knownAppIds.TryGetValue(instrumentationKey, out correlationId);
```

Comment:
There are many ways an instrumentation key might be set. One of these ways is using a custom telemetry initializer. Hence the right way to ascertain instrumentation key is from the telemetry object before it is sent. It might be different from what's in the config. 

This cache saves the app ids against instrumentation keys (can be more than one) and makes sure we don't always call the breeze endpoint.